File: Sources/Olaf/Network/PendingRequestRegistry.swift
import Foundation

/// An in-flight (not yet completed) network capture.
public struct PendingNetworkRequest: Identifiable, Sendable {
    public let id: UUID
    public let method: String
    public let url: String
    public let startDate: Date

    /// Elapsed time since start (seconds).
    public var elapsedSeconds: Int {
        max(0, Int(Date().timeIntervalSince(startDate)))
    }
}

/// Lock-protected registry of in-flight captures. `OlafURLProtocol` registers a request when it
/// starts, and drops it on completion (success/error/cancel). Feeds the viewer's "Active requests"
/// section — a snapshot is read on a 1s `TimelineView` tick; no separate broadcast mechanism needed.
final class PendingRequestRegistry: @unchecked Sendable {

    static let shared = PendingRequestRegistry()

    private let lock = NSLock()
    private var items: [UUID: PendingNetworkRequest] = [:]

    func register(method: String, url: String) -> UUID {
        let request = PendingNetworkRequest(id: UUID(), method: method, url: url, startDate: Date())
        lock.lock(); items[request.id] = request; lock.unlock()
        return request.id
    }

    func unregister(_ id: UUID) {
        lock.lock(); items[id] = nil; lock.unlock()
    }

    /// Snapshot sorted by start time (oldest first).
    var snapshot: [PendingNetworkRequest] {
        lock.lock(); defer { lock.unlock() }
        return items.values.sorted { $0.startDate < $1.startDate }
    }

    /// For test isolation: drops all entries.
    func removeAll() {
        lock.lock(); items.removeAll(); lock.unlock()
    }
}

File: Android/olaf/src/main/java/com/olaf/network/PendingRequestRegistry.kt
package com.olaf.network

import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/** An in-flight (not yet completed) network capture. */
data class PendingNetworkRequest(
    val id: String,
    val method: String,
    val url: String,
    val startedAtMillis: Long
) {
    /** Seconds elapsed since the request started. */
    val elapsedSeconds: Long
        get() = ((System.currentTimeMillis() - startedAtMillis) / 1000).coerceAtLeast(0)
}

/**
 * Registry of in-flight captures. [OlafInterceptor] registers a request when it starts and drops
 * it on completion (success, failure or cancellation). This feeds the viewer's "Active requests"
 * section, which polls a snapshot on a timer — no separate broadcast is needed.
 */
internal object PendingRequestRegistry {

    private val items = ConcurrentHashMap<String, PendingNetworkRequest>()

    fun register(method: String, url: String): String {
        val id = UUID.randomUUID().toString()
        items[id] = PendingNetworkRequest(id, method, url, System.currentTimeMillis())
        return id
    }

    fun unregister(id: String) {
        items.remove(id)
    }

    /** Snapshot ordered by start time, oldest first. */
    fun snapshot(): List<PendingNetworkRequest> = items.values.sortedBy { it.startedAtMillis }

    /** For test isolation. */
    fun removeAll() {
        items.clear()
    }
}
